Apply sst file manager in myrocks clone to add slow-rm during checkpoints removal

sunshine-Chun opened the pull request:
This diff makes the following two changes:
1. Clone use rocksdb::DestroyDB() to clean the checkpoint. However, we didn't pass the sst file manager so didn't make use of the slow-rm feature provided by sst file manager. This will lead up to a discard spike when rolling checkpoints. This change pass the sst file manager when cleaning checkpoints.
2. During slow-rm sst files, the checkpoint directory may not be deleted immediately. In the rolling checkpoint function call, we will first delete the checkpoint directory, then create a new one with the same name. Since we are slow removing the sst files, it's possible that when we create the new checkpoints, the old directory hasn't been deleted yet. This will cause a 'directory exist' error when creating a new checkpoint directory. In this change, we add an additional suffix to differentiate rolling checkpoints directory to avoid this error.

sunshine-Chun committed:
Apply sst file manager in myrocks clone to add slow-rm during checkpoints removal

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: https://phabricator.intern.facebook.com/D50758286

laurynas-biveinis commented:
@sunshine-Chun, I am getting linker errors now:
```
ld: Undefined symbols:
  myrocks::rocksdb_remove_checkpoint(char const*), referenced from:
      myrocks::rocksdb_close_connection(handlerton*, THD*) in librocksdb_se.a[2](ha_rocksdb.cc.o)
      myrocks::rocksdb_create_temporary_checkpoint_validate(THD*, SYS_VAR*, void*, st_mysql_value*) in librocksdb_se.a[2](ha_rocksdb.cc.o)
      (anonymous namespace)::rdb_checkpoint::cleanup() in librocksdb_se.a[25](donor.cc.o)

```

sunshine-Chun commented:
@laurynas-biveinis Did you hit it during clone? Can you elaborate how you hit this error?

laurynas-biveinis commented:
> @laurynas-biveinis Did you hit it during clone? Can you elaborate how you hit this error?

It's a linker error during server build

laurynas-biveinis commented on storage/rocksdb/clone/donor.cc:
```diff
@@ -102,10 +105,14 @@ class [[nodiscard]] rdb_checkpoint final {
 
   static std::atomic<std::uint64_t> m_next_id;
 
+  std::atomic<std::uint64_t> m_next_sub_id{1};
```
This should not be atomic. It is private to a single clone session, and all its accesses happen with `m_donor_mutex` locked.

sunshine-Chun replied:
resolved

laurynas-biveinis commented on storage/rocksdb/clone/donor.cc:
```diff
@@ -68,7 +68,9 @@ class [[nodiscard]] rdb_checkpoint final {
   [[nodiscard]] int init() {
     assert(!m_active);
-    const auto result = myrocks::rocksdb_create_checkpoint(m_dir.c_str());
+    m_full_dir = make_dir_full_name(
```
It might be confusing to have `m_dir` and `m_full_dir` and `make_dir_name` and `make_dir_full_name`. After this patch, the dir name is the one that has all the suffixes, and the previous dir name is a prefix. Please rename `make_dir_name` to `make_dir_prefix_name`, `make_dir_full_name` to `make_dir_name`, and member variables too.

sunshine-Chun replied:
resolved

laurynas-biveinis commented on storage/rocksdb/clone/donor.cc:
```diff
@@ -120,6 +127,19 @@ class [[nodiscard]] rdb_checkpoint final {
     return result;
   }
+
+  [[nodiscard]] static std::string make_dir_full_name(std::string base_dir,
```
```suggestion
  [[nodiscard]] static std::string make_dir_full_name(const std::string &dir_name_prefix,
```

sunshine-Chun replied:
resloved

laurynas-biveinis commented on storage/rocksdb/clone/donor.cc:
```diff
@@ -68,7 +68,9 @@ class [[nodiscard]] rdb_checkpoint final {
     assert(!m_active);
-    const auto result = myrocks::rocksdb_create_checkpoint(m_dir.c_str());
+    m_full_dir = make_dir_full_name(
+        m_dir.c_str(), m_next_sub_id.fetch_add(1, std::memory_order_relaxed));
```
No need to take a C string out of a C++ string to convert it to a C++ string again:
```suggestion
        m_dir, m_next_sub_id.fetch_add(1, std::memory_order_relaxed));
```

sunshine-Chun replied:
resolved.

laurynas-biveinis commented on storage/rocksdb/ha_rocksdb.cc:
```diff
@@ -1325,6 +1311,27 @@ static void rocksdb_set_io_write_timeout(
+           rocksdb_hton_name);
+  return HA_EXIT_FAILURE;
+}
+
```
So the reason for my linker error is that this function now fell into a `#ifndef __APPLE__` block, please pull outside

sunshine-Chun replied:
resolved

laurynas-biveinis commented on storage/rocksdb/ha_rocksdb.cc:
```diff
@@ -1325,6 +1311,27 @@ static void rocksdb_set_io_write_timeout(
+                  "deleting temporary checkpoint in directory : %s\n",
+                  checkpoint_dir.c_str());
+
+  std::string trash_dir = std::string(rocksdb_datadir) + "/trash";
```
IMHO this is important to deduplicate with the other trash dir occurence in `rocksdb_init_internal`, please extract a helper to get the trash dir.

```suggestion
  const auto trash_dir = std::string(rocksdb_datadir) + "/trash";
```

laurynas-biveinis commented on storage/rocksdb/ha_rocksdb.cc:
```diff
@@ -1325,6 +1311,27 @@ static void rocksdb_set_io_write_timeout(
+                  checkpoint_dir.c_str());
+
+  std::string trash_dir = std::string(rocksdb_datadir) + "/trash";
+  rocksdb::Options op = rocksdb::Options();
```
```suggestion
  auto op = rocksdb::Options();
```

sunshine-Chun replied:
resolved

laurynas-biveinis commented on storage/rocksdb/ha_rocksdb.cc:
```diff
@@ -1325,6 +1311,27 @@ static void rocksdb_set_io_write_timeout(
+
+  std::string trash_dir = std::string(rocksdb_datadir) + "/trash";
+  rocksdb::Options op = rocksdb::Options();
+  op.sst_file_manager.reset(NewSstFileManager(
```
Not an expert in SST file manager so open questions:
1) can we save reuse the existing manager from `rocksdb_init_internal` as singleton?
2) if no, is it really correct to pass `delete_existing_trash = true` to this one? Wouldn't this be stepping on the toes of the other SST file manager instances?

sunshine-Chun replied:
Checked with Rocksdb team. The trash_dir has actually been deprecated. It's the old way of managing trash files. Nowadays Rocksdb team manage them by renaming the sst files. I will pass nullptr to trash dir and set the delete_existing_trash to false since we are not using them.

laurynas-biveinis replied:
> Checked with Rocksdb team. The trash_dir has actually been deprecated. It's the old way of managing trash files. Nowadays Rocksdb team manage them by renaming the sst files. I will pass nullptr to trash dir and set the delete_existing_trash to false since we are not using them.

I see, thank you. Should we convert the existing `rocksdb_init_internal` usage as well, in this or a follow-up PR? And can we reuse that same manager object?

sunshine-Chun replied:
Yes we can convert the existing rocksdb_init_internal usage in a follow-up PR. 

 >And can we reuse that same manager object?

I am open for suggestion. My original thought is that we only have two places use sst_file_manager. Does it worth to maintain a static variable for the usage of these two places?

laurynas-biveinis replied:
> I am open for suggestion. My original thought is that we only have two places use sst_file_manager. Does it worth to maintain a static variable for the usage of these two places?

I am not sure neither. Clone can run many times, and rotate the checkpoint many times too, so dynamically it could be much more than two places. Also the init options are repeated, and probably should be in sync after the `rocksdb_init_internal` follow-up. A static variable for this would be closely related to `rdb` static var, which looks fine to me.

laurynas-biveinis replied:
Forgot one more thing, is reusing the same object even safe in the first place?

sunshine-Chun committed:
Address comments

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Fix comment

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

laurynas-biveinis commented:
@sunshine-Chun , since the original PR has been merged already, I believe a new PR has to be made instead of pushing new commits to this one

sunshine-Chun commented:
create a new pr: Address comments for 'Apply sst file manager in myrocks clone' #1388
